src/main/java/picard/illumina/IlluminaBasecallsToFastq.java

```diff
@@ -486,17 +529,48 @@ public void write(final ClusterData rec) {
                         throw new IllegalStateException("Read type other than T/B/M encountered.");
                 }
 
-                writeSingle(out, name, read);
+                writeSingle(out, name, read, templateIndex);
             }
         }
 
         /**
          * Writes out a single read to a single FASTQ's output stream.
          */
-        private void writeSingle(final OutputStream out, final String name, final ReadData read) {
+        private void writeSingle(final OutputStream out, final String name, final ReadData read, int templateIndex) {
             try {
-                final byte[] bases = read.getBases();
-                final byte[] quals = read.getQualities();
+                byte[] bases = read.getBases();
+                byte[] quals = read.getQualities();
+
+                if (trimmingQuality != null) {
+                    int index = TrimmingUtil.findQualityTrimPoint(quals, trimmingQuality);
+                    // Don't write zero length reads if the entire read is trimmed
+                    if (index == 0) {
+                        return;
+                    }
+                    quals = Arrays.copyOfRange(quals, 0, index);
+                    bases = Arrays.copyOfRange(bases, 0, index);
+                }
+
+                if (adapterMarker != null) {
+                    AdapterPair adapterPair = adapterMarker.findAdapterPairForSingleRead(bases, templateIndex);
+                    if (adapterPair != null) {
+                        int index = bases.length - 1;
+                        if (templateIndex == 1) {
+                            index = ClippingUtility.findIndexOfClipSequence(
+                                    bases, adapterPair.get3PrimeAdapterBytes(), ClippingUtility.MIN_MATCH_BASES, ClippingUtility.MAX_ERROR_RATE);
+                        } else if (templateIndex == 2) {
+                            index = ClippingUtility.findIndexOfClipSequence(
+                                    bases, adapterPair.get5PrimeAdapterBytesInReadOrder(), ClippingUtility.MIN_MATCH_BASES, ClippingUtility.MAX_ERROR_RATE);
+                        }
```

Comment:
I had to go look in ClippingUtility and it's also like this, which is crazy!  So we search the read once in order to figure out _which_ adapter pair it is, and then if we got a match, we go back and do it _again_ to re-compute where it is?  Ugh.

Could we maybe add methods to AdapterMarker/ClippingUtiliity that return a tuple of `<AdapterPair, Integer>` to give you both at the same time and avoid re-matching?